### PR TITLE
YOLO Changes

### PR DESCRIPTION
**Pull request by LilAttacker**
# YOLO Changes

## Description

This is a draft and will take some time for this to be eventually merged.

Currently, the only is to the output of the YOLO model. Now it gives cropped forms of the images for the objects its found along with text boxes describing them as well.

Also changed to provide a json output when called so you can actually look at each object it found more in depth.

This will be expanded later when closer to proper plan for YOLO implementation.

## Type of Change

- [ ] Bugfix
- [ ] New Feature
- [X] Documentation Update
- [X] Other (please describe): Slightly expands output of YOLO

## Checklist

- [X] I have read the contributing guidelines.
- [X] I have followed the code style of this project.
- [X] I have added tests to cover my changes.
- [X] I have updated the documentation (if applicable).
- [X] My changes do not generate new warnings.

## Related Issue

#32 

## Additional Notes

I plan to eventually make this into a proper PR with final implementation. But for now its a draft to show what is currently the progress on it.

If its desired, this is workable. So if someone wants to test something on this quick, we can merge it and there shouldn't be any issues.

### CODE REVIEW

**Comment by LilAttacker** [comment]
Added a more "official" function for YOLO to be called from internally. Currently doesn't work and is untested as I'm awaiting some design choices.

**Comment by LilAttacker** [comment]
New output currently outputs in a more EasyOCR friendly format (#34).
Current iteration doesn't make images anywhere (#48).
New function looks at specific file path for dynamic image input (#49).

**Comment by LilAttacker** [comment]
Alright, I've updated to using a segmentation model for image recognition.

It's still not great, but it works for close up images of shelves for the most part. Just need to know where images will be temporarily stored and make some tests, then this would basically be ready.

**Comment by LilAttacker** [comment]
This is almost finished.

Latest Changes Include
- Input for YOLO is now a tensor image.
- Added exception testing to ensure tensor image is compatible with YOLO.
- Added tests to cover the new YOLO function.
- Added a context manager to make it more consistent when testing or using the new YOLO function.
- Model does feel more accurate when using the image matrix instead of jpeg.
- Added more example images.

File: server/tabby_server/vision/image_labelling.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+# Expected Image Size for Model to Use
+EXPECTED_SIZE = (640, 640)
+
+# Temporarily Change Working Directory
+with temp_dir("tabby_server/vision/"):
+    # Load pretrained model from other directory.
+    model = YOLO("shelf_yolo.pt")
+
+
+def find_books(tensor_image=None) -> dict[str, list]:
```
Comment (ederic-oytas) [inline server/tabby_server/vision/image_labelling.py:46]: The return type annotation is inconsistent with the documentation. `list of dicts` versus `dict[str, list]`

**Review by ederic-oytas — CHANGES_REQUESTED** [review]
Hi Daniel, this looks good. The only thing I want you to correct is the type annotation/documentation for `find_books()`, which are inconsistent with each other. Style-wise, I have no complaints.
[X] Feature complete
[X] New test coverage, all tests pass
[X] Follows unified style
[X] Has documentation

File: server/tabby_server/vision/image_labelling.py (hunk 1 above)
Comment (LilAttacker) [reply server/tabby_server/vision/image_labelling.py]: Fixed.

**Review by David1000734 — COMMENTED** [review]
Looks good, although I'm wondering is it possible for you to do some type of download using a link from google drive for the images as to avoid putting too much images into the server and increasing it's size?

File: server/tabby_server/vision/image_labelling.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+        yield
+    finally:
+        os.chdir(original_dir)
+
+
+# Expected Image Size for Model to Use
+EXPECTED_SIZE = (640, 640)
+
+# Temporarily Change Working Directory
+with temp_dir("tabby_server/vision/"):
```
Comment (ederic-oytas) [inline server/tabby_server/vision/image_labelling.py:41]: Is temporarily changing the directory really necessary? Can you just load it by using `YOLO("tabby_server/vision/shelf_yolo.pt")`?

Comment (David1000734) [reply server/tabby_server/vision/image_labelling.py:41]: Add a TODO: comment on this so we are reminded to check on this for a better implementation later.

**Review by David1000734 — APPROVED** [review]

File: server/tabby_server/vision/yolo_test.py [hunk 3]
```diff
@@ -16,53 +17,104 @@
...
+    import torch
+
+    tensor_image = (
+        torch.from_numpy(normalized_image)
+        .float()
+        .permute(2, 0, 1)
+        .unsqueeze(0)
+    )
+
+    # ai-gen end
```
Comment (LilAttacker) [inline server/tabby_server/vision/yolo_test.py:47]: Here is the ChatGPT conversation that made this: https://chatgpt.com/share/67359936-d608-800f-a86e-3eab0a9e9340

File: server/tabby_server/vision/image_labelling.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+        and tensor_image.shape[2:] == EXPECTED_SIZE
+        and tensor_image.dtype == torch.float32
+        and tensor_image.max() <= 1.0
+        and tensor_image.min() >= 0.0
+    ):
+        return [{"bad_tensor": [1]}]
+    # tensor_image = tensor_image.float()   # Converts tensor to float32.
+    # tensor_image = tensor_image / 255.0   # Normalizes tensor image.
+
+    # ai-gen end
```
Comment (LilAttacker) [inline server/tabby_server/vision/image_labelling.py:156]: ChatGPT conversation: https://chatgpt.com/share/67359936-d608-800f-a86e-3eab0a9e9340

File: server/tests/test_yolo.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    bad_image = torch.randint(0, 256, (1, 3, 640, 640), dtype=torch.uint8)
+    result = find_books(bad_image)
+    assert result is not None and "bad_tensor" in result[0]
+
+    # Incorrect Normalization.
+    bad_image = bad_image.float()
+    result = find_books(bad_image)
+    assert result is not None and "bad_tensor" in result[0]
+
+    # ai-gen end
```
Comment (LilAttacker) [inline server/tests/test_yolo.py:50]: ChatGPT conversation: https://chatgpt.com/share/67359936-d608-800f-a86e-3eab0a9e9340

**Pull request merged**